build-tools/packages/build-tools/src/common/tsCompile.ts

```diff
@@ -8,27 +8,57 @@ import type * as tsTypes from "typescript";
 import { defaultLogger } from "./logging.js";
 import { getTscUtils, normalizeSlashes } from "./tscUtils.js";
 
-export function tsCompile({
-	command,
-	cwd,
-	packageJsonTypeOverride,
-}: {
+interface TsCompileOptions {
+	/**
+	 * Complete tsc command line
+	 */
 	command: string;
+	/**
+	 * Working directory
+	 */
 	cwd: string;
+	/**
+	 * When specified local package.json will be interpretted as
+	 * having "types" property set to this value.
+	 */
 	packageJsonTypeOverride?: "commonjs" | "module";
-}): number {
+}
+
+/**
+ * Executes given tsc command line.
+ * If command line includes `--watch`, an error will be thrown.
+ *
+ * @param options - {@link TsCompileOptions}
+ * @returns numeric exit code
+ */
+export function tsCompile(options: TsCompileOptions): number;
+/**
+ * Executes given tsc command line that may include `--watch`.
+ *
+ * @param options - {@link TsCompileOptions}
+ * @returns numeric exit code when completing immediately or undefined when watch has started
+ */
+export function tsCompile(
+	options: TsCompileOptions,
+	allowWatch: "allow-watch",
```

Comment:
What's the motivation for doing this via overloads and a string parameter vs just taking an optional `allowWatch` boolean?